### PR TITLE
Add script to run binaries in QEMU with LAM.

### PR DESCRIPTION
**Pull request by morehouse**
Can be used in combination with LLVM's COMPILER_RT_EMULATOR CMake
variable and check-hwasan-lam to test LAM functionality in QEMU.

### CODE REVIEW

File: hwaddress-sanitizer/run_in_qemu_with_lam.sh [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  (force_kill_qemu_after_timeout) &
+  if kill "${QEMU_PID}"; then
+    echo "Waiting for QEMU to shutdown..."
+    wait "${QEMU_PID}" &>/dev/null || true
+  fi
+
+  echo "Done!"
+}
+
+function wait_pipe_for_match {
```
Comment (vitalybuka) [inline hwaddress-sanitizer/run_in_qemu_with_lam.sh:71]: why pipes and not just ssh \<command\>?
If this is to keep just a single ssh connection then "-M" should be enough

File: hwaddress-sanitizer/run_in_qemu_with_lam.sh [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+QEMU_PID=""
+SSH_PORT=""
+
+function force_kill_qemu_after_timeout {
+  sleep "${QEMU_FORCE_KILL_TIMEOUT}"
+  kill -9 "${QEMU_PID}" &>/dev/null || true
+}
+
+function on_exit {
+  (force_kill_qemu_after_timeout) &
```
Comment (vitalybuka) [inline hwaddress-sanitizer/run_in_qemu_with_lam.sh:62]: Why this not just 
force_kill_qemu_after_timeout &

File: hwaddress-sanitizer/run_in_qemu_with_lam.sh [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+  echo "Waiting for QEMU to login..."
+  wait_pipe_for_match "debian login: root (automatic login)"
+}
+
+function copy_to_qemu {
+  local local_file="${1}"
+  local qemu_dir="${2}"
+
+  scp -P "${SSH_PORT}" -o "StrictHostKeyChecking=no" \
```
Comment (vitalybuka) [inline hwaddress-sanitizer/run_in_qemu_with_lam.sh:115]: As future improvement maybe try:
HOST probably can try to sshfs empty dir from VM and do build/tests build part will be obviously slower but tests probably faster
We can build/test compiler-rt as standalone, so this dir will not be used for clang/llvm build

File: hwaddress-sanitizer/run_in_qemu_with_lam.sh (hunk 1 above)
Comment (morehouse) [reply hwaddress-sanitizer/run_in_qemu_with_lam.sh:71]: I still don't fully understand what's going on, but it seems that having too many SSH commands running in parallel is triggering the failures.  If I reduce the number of llvm-lit workers, I get fewer failures.

I don't have this issue using the pipe method implemented here.

Comment (vitalybuka) [reply hwaddress-sanitizer/run_in_qemu_with_lam.sh:71]: Even with -M? For -M you need to keep the first ssh connection open.

Comment (morehouse) [reply hwaddress-sanitizer/run_in_qemu_with_lam.sh]: Thanks.  Indeed `-M` seems to be the trick.  Using it, I actually get faster test executions than using pipes, *and* the strange test failures go away.

I've uploaded a new diff with this approach.

**Review by vitalybuka — APPROVED** [review]

**Pull request merged**